File: 1458-sort-integers-by-the-number-of-1-bits/sort-integers-by-the-number-of-1-bits.java
class Solution {
    public int[] sortByBits(int[] arr) {
        Integer[] integerArr = new Integer[arr.length];
        for(int i =0; i<arr.length; i++){
            integerArr[i] = arr[i];
        }
        Arrays.sort(integerArr, (a,b) -> {
            int bitcountA = Integer.bitCount(a);
            int bitcountB = Integer.bitCount(b);

            if(bitcountA != bitcountB){
                return bitcountA - bitcountB;
            }
            return a-b;
        });
        for(int i = 0; i< arr.length; i++){
            arr[i] = integerArr[i];
        }
        return arr;
    }
}

File: 1458-sort-integers-by-the-number-of-1-bits/sort-integers-by-the-number-of-1-bits.py
class Solution:
    def sortByBits(self, arr: list[int]) -> list[int]:
        # Sort based on a tuple: (bit_count, original_value)
        # Python handles the tie-breaking automatically!
        arr.sort(key=lambda x: (bin(x).count('1'), x))
        return arr